src/lib.rs

```diff
@@ -1,6 +1,164 @@
+//!
+//! ### From text Ion
+//! The [Element::read_one] method will parse the provided data and requires that it contain exactly
+//! one Ion value.
+//! ```
+//! # use ion_rs::IonResult;
+//! # fn main() -> IonResult<()> {
+//! use ion_rs::element::owned::Element;
+//! use ion_rs::IonType;
+//! let ion_data = "[1, 2, 3]";
+//! let element = Element::read_one(ion_data)?;
+//! assert_eq!(element.ion_type(), IonType::List);
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! [Element::read_all] will read any number of Ion values and return them as a `Vec<Element>`.
+//!
+//! [Element::read_first] will read the first Ion value without requiring that the stream have
+//! exactly one value.
+//!
+//! ### From a Rust value
+//! Most Rust primitives implement `Into<Element>`, allowing them to be converted to an Ion [Element]
+//! directly.
+//! ```
+//! # use ion_rs::IonResult;
+//! # fn main() -> IonResult<()> {
+//! use ion_rs::element::owned::Element;
+//!
+//! let int: Element = 5.into();
+//! assert_eq!(Element::read_one("5")?, int);
+//!
+//! let boolean: Element = true.into();
+//! assert_eq!(Element::read_one("true")?, boolean);
+//!
+//! let string: Element = "hello".into();
+//! assert_eq!(Element::read_one("\"hello\"")?, string);
+//!
+//! let ion_version_marker: &[u8] = &[0xE0, 0x01, 0x00, 0xEA]; // Ion 1.0 version marker
+//! let blob: Element = ion_version_marker.into();
+//! assert_eq!(Element::read_one("{{4AEA6g==}}")?, blob);
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! ### Using macros
+//!
+//! When constructing a container [Element], you can use the [`ion_list!`], [`ion_sexp!`],
+//! and [`ion_struct!`] macros.
+//!
+//! ```
+//! # use ion_rs::IonResult;
+//! # fn main() -> IonResult<()> {
+//! use ion_rs::element::owned::Element;
+//! use ion_rs::{ion_list, ion_sexp, ion_struct};
+//!
+//! // Variable names are allowed
+//! let six = 6i64;
+//! let list: Element = ion_list! [true, six, "foo"].into();
+//! assert_eq!(Element::read_one("[true, 6, \"foo\"]")?, list);
+//!
+//! // Nested use of macros is allowed
+//! // Notice that ion_sexp! uses ()s without commas
+//! let sexp: Element = ion_sexp! (true six ion_list!["foo", "bar"]).into();
+//! assert_eq!(Element::read_one("(true 6 [\"foo\", \"bar\"])")?, sexp);
+//!
+//! let field_name = "bar";
+//! let struct_: Element = ion_struct! {
+//!   "foo": six,
+//!   field_name: false
+//! }.into();
+//! assert_eq!(Element::read_one("{foo: 6, bar: false}")?, struct_);
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! ### From a stream
+//!
+//! ```no_run
+//! # use ion_rs::IonResult;
+//! # fn main() -> IonResult<()> {
+//! use std::fs::File;
+//! use ion_rs::element::owned::Element;
+//! use ion_rs::element::reader::ElementReader;
+//! use ion_rs::ReaderBuilder;
+//! let ion_file = File::open("/foo/bar/baz.ion").unwrap();
+//! let mut reader = ReaderBuilder::default().build(ion_file)?;
+//! // A simple pretty-printer
+//! for element in reader.elements() {
+//!     println!("{}", element?)
```

Comment:
Should we add a sample file content and output for this example?

Reply:
It does stand out as the only example without an `assert_eq!`, but I figured seeing how to access the elements in a file would be enough for folks to use the other tests' examples. I'm going to leave this as-is for the moment to get v0.16.0 out, but I'm open to changing it in the future.